### DISCUSSION TITLE
Usage of dictionaries / the value map from the script context

### DISCUSSION THREAD

**Discussion by nielstron** [OpShin/opshin#38] [Show and tell]
The script context provides a dictionary/map from policy id -> tokenname -> value being contained in the inputs and outputs of the transaction being validated.

This thread is used to keep track of questions popping up regarding its use.

**Comment by nielstron** [comment]
*Q:* does value.get require an extra parameter here ->
```python
def entry_policy_present(
    txins: List[TxInInfo]
) -> Bool:
    res = False
    for txi in txins:
        if txi.resolved.value.get(token.policy_id) == ENTRY-POLICY:
            res = True
    return res
```

*A:* Yes! As opposed to the way it is done in python, you _always_ need to supply a default value when accessing elements in a dictionary in eopsin. The reason is that _None_ is not always of the correct type, so always supply a default value of the correct type. In this case the type is a dictionary from bytestring to integers, you can for example supply `{b"":0}`.

**Comment by while0x1** [comment]
so in that instance to check for the existance of ENTRY-POLICY the correct should be txo.value.get(token.policy_id, {b"": 0}) ?

**Reply by nielstron** [reply]
You need to check that there is any token being spent with the ENTRY-POLICY id. txo.value gives you a map from policy id to a map from tokenname to how much of it is being spent. Make sure that your python code is valid python code first and `eopsin compile` your program to get additional hints from the compiler!

**Comment by nielstron** [comment]
*Q:* for getting the ada being spent by the contract ->
```python
own_utxo = own_spent_utxo(ctx.tx_info.inputs, purpose)
spendingvalue = own_utxo.value.get...
```

*A:* In this case, we need to access the policy id and token name of ADA. It is defined in the plutus docs that ADA/Lovelace always has the empty bytestring as policy id and token name. So we use `own_utxo.value.get(b"", {b"":0}).get(b"",0)`. Note how the default value has to match the expected return type!
